exporter.go

```diff
@@ -351,105 +338,159 @@ func parseDogStatsDTagsToLabels(component string) map[string]string {
-					switch component[0] {
-					case '@':
-						if statType != "c" && statType != "ms" {
-							log.Errorln("Illegal sampling factor for non-counter metric on line", line)
-							networkStats.WithLabelValues("illegal_sample_factor").Inc()
-							continue
-						}
-						samplingFactor, err = strconv.ParseFloat(component[1:], 64)
-						if err != nil {
-							log.Errorf("Invalid sampling factor %s on line %s", component[1:], line)
-							networkStats.WithLabelValues("invalid_sample_factor").Inc()
-						}
-						if samplingFactor == 0 {
-							samplingFactor = 1
-						}
-
-						if statType == "c" {
-							value /= samplingFactor
-						} else if statType == "ms" {
-							multiplyEvents = int(1 / samplingFactor)
-						}
-					case '#':
-						labels = parseDogStatsDTagsToLabels(component)
-					default:
-						log.Errorf("Invalid sampling factor or tag section %s on line %s", components[2], line)
+					samplingFactor, err = strconv.ParseFloat(component[1:], 64)
+					if err != nil {
+						log.Errorf("Invalid sampling factor %s on line %s", component[1:], line)
 						networkStats.WithLabelValues("invalid_sample_factor").Inc()
-						continue
 					}
-				}
-			}
+					if samplingFactor == 0 {
+						samplingFactor = 1
+					}
 
-			for i := 0; i < multiplyEvents; i++ {
-				event, err := buildEvent(statType, metric, value, relative, labels)
-				if err != nil {
-					log.Errorf("Error building event on line %s: %s", line, err)
-					networkStats.WithLabelValues("illegal_event").Inc()
+					if statType == "c" {
+						value /= samplingFactor
+					} else if statType == "ms" {
+						multiplyEvents = int(1 / samplingFactor)
+					}
+				case '#':
+					labels = parseDogStatsDTagsToLabels(component)
+				default:
+					log.Errorf("Invalid sampling factor or tag section %s on line %s", components[2], line)
+					networkStats.WithLabelValues("invalid_sample_factor").Inc()
 					continue
 				}
-				events = append(events, event)
 			}
-			networkStats.WithLabelValues("legal").Inc()
 		}
+
+		for i := 0; i < multiplyEvents; i++ {
+			event, err := buildEvent(statType, metric, value, relative, labels)
+			if err != nil {
+				log.Errorf("Error building event on line %s: %s", line, err)
+				networkStats.WithLabelValues("illegal_event").Inc()
+				continue
+			}
+			events = append(events, event)
+		}
+		networkStats.WithLabelValues("legal").Inc()
+	}
+	return
+}
+
+type StatsDUDPListener struct {
+	conn *net.UDPConn
+}
+
+func (l *StatsDUDPListener) Listen(e chan<- Events) {
+	buf := make([]byte, 65535)
+	for {
+		n, _, err := l.conn.ReadFromUDP(buf)
+		if err != nil {
+			log.Fatal(err)
+		}
+		l.handlePacket(buf[0:n], e)
+	}
+}
+
+func (l *StatsDUDPListener) handlePacket(packet []byte, e chan<- Events) {
+	lines := strings.Split(string(packet), "\n")
+	events := Events{}
+	for _, line := range lines {
+		events = append(events, lineToEvents(line)...)
 	}
 	e <- events
 }
+
+type StatsDTCPListener struct {
```

Comment:
We don't *need* a separate listener -- assuming we have a type switch. It seems like it'd be *less* code-- but this definitely isn't wrong ;)